Revert some deprecations, following asyncio changes

takluyver opened the pull request:
This is an attempt to tackle #3216 - at least the straightforward parts of it.

- Remove deprecation warnings and notices for ioloop `make_current` and `clear_current` methods, and the `make_current` constructor parameter.
- Remove deprecation notices on `AsyncTestCase` and `AsyncHTTPTestCase`
- Remove the mention of Python changes from the TCPServer `bind` & `start` deprecation notices (but they're still deprecated)

I've assumed that removing the deprecation notices is sufficient documentation - we could add notes saying that they were deprecated and then undeprecated again, but that feels like unnecessary clutter.

Still to figure out:

- [ ] `IOLoop.current()` has a deprecation notice for using it without a running asyncio loop. Is this also un-deprecated? Does it need some code change to handle when there's no running loop nor a default loop specifically set in asyncio?
- [ ] Are some changes needed to this code in AsyncTestCase? https://github.com/tornadoweb/tornado/blob/59536016e37704a1c47a3a0d262f46ceb6132c06/tornado/testing.py#L204-L238
- [x] `AnyThreadEventLoopPolicy` remains deprecated (as mentioned in #3216), but is any change needed to its deprecation notice? https://github.com/tornadoweb/tornado/blob/59536016e37704a1c47a3a0d262f46ceb6132c06/tornado/platform/asyncio.py#L423-L431

takluyver commented:
I think the CI failure is something random unrelated to these changes. It's failing in a test about subprocess handling, only on Python 3.10, and only with `LANG=en_US.utf-8`. The changes so far shouldn't make any functional difference, except for getting rid of some warnings.

bdarnell commented:
Thanks!

These changes look good, but to be sure I want to wait until we can test with the next alpha of python 3.12 (due next week) before merging.

The CI issue is a known flaky test which I've just written up as #3225. We should probably just skip that test since it's not helpful to trip over it every once in a while.

`IOLoop.current`: Yes, I think we'll need some code changes there to keep the current semantics. Probably something in the `except` block after `get_event_loop` that creates a new one if it doesn't exist. Or we could leave this as-is and adopt asyncio's stance against implicit creation of event loops. That would mean in practice that most applications would still want to adopt `asyncio.run`, so I think it's more important for us to keep backwards compatibility than to track asyncio on this point.

In the test classes, I think we can just revert that whole block from https://github.com/tornadoweb/tornado/pull/3097/files#diff-0c44c25deecc01d7f48e0fd36005e637bb772273e31377f21fb16690c285dc4b - the old code should work again. But this is something we'd need to test with the new alpha.

AnyThreadEventLoopPolicy's deprecation comment is still valid. In general I think we're probably fine with removing the deprecation notices entirely. We'll explain the undeprecations in the 6.3 release notes and amend the ones from 6.2 to note what's changed again.

takluyver commented:
Thanks! No problem about waiting for the next 3.12 alpha. From [this discussion](https://github.com/python/cpython/issues/100160), it looks like the behaviour changes (not creating a default event loop for the main thread) are being delayed to 3.14, because the deprecation warnings weren't thorough enough.

I've changed `IOLoop.current()` to create a new event loop if necessary, but only on the main thread. It doesn't seem to be documented that this behaves differently depending on the thread, but there is a test that fails if it doesn't (part of `AnyThreadEventLoopPolicyTest` checks the behaviour with the default policy). Would you rather I changed the behaviour and the test so `IOLoop.current(instance=True)` can create a loop on any thread?

> In the test classes, I think we can just revert that whole block from https://github.com/tornadoweb/tornado/pull/3097/files#diff-0c44c25deecc01d7f48e0fd36005e637bb772273e31377f21fb16690c285dc4b - the old code should work again. 

I'm still a bit confused about this. The comment added in that PR says "packages such as pytest-asyncio (as of version 0.16.0) still rely on the implicit current event loop and we want to be compatible with them". But I think reverting to the code before that will always create a new event loop (calling `IOLoop()`) and set it as the current one.

It's unfortunate that there appears to be no way of getting the future (Python 3.14) behaviour now - there's no public attribute or method to check if the current thread already has an event loop without potentially creating a new one.

takluyver commented:
OK, I've rebased, and all the tests seem to be passing with the new 3.12 alpha.

bdarnell commented on tornado/ioloop.py:
```diff
@@ -254,10 +254,13 @@ def current(instance: bool = True) -> Optional["IOLoop"]:  # noqa: F811
         try:
             loop = asyncio.get_event_loop()
-        except (RuntimeError, AssertionError):
+        except RuntimeError:
```
Are you removing `AssertionError` here based on [this comment](https://github.com/tornadoweb/tornado/blob/9e9043e3cbf4605b272f74ede2c8e6a4d71049c1/tornado/platform/asyncio.py#L447) that says `AssertionError` was only used in Python 3.4?

takluyver replied:
Yup, that's what I went on.

bdarnell commented on tornado/ioloop.py:
```diff
@@ -257,7 +258,9 @@ def current(instance: bool = True) -> Optional["IOLoop"]:  # noqa: F811
             if not instance:
                 return None
-            # Create a new asyncio event loop for this thread.
+            if threading.current_thread() is not threading.main_thread():
```
I don't think we want the main-thread-only behavior here. Tornado has historically allowed event loop creation on any thread (that's what AnyThreadEventLoopPolicy is for), so it's a nice side effect that we can bring back the "any thread" behavior by default at the same time that AnyThreadEventLoopPolicy is being deprecated. 

Just remove this commit.

takluyver replied:
OK, I've removed that commit, and adjusted `AnyThreadEventLoopPolicyTest` to match.

bdarnell reviewed (commented):
> I'm still a bit confused about this. The comment added in that PR says "packages such as pytest-asyncio (as of version 0.16.0) still rely on the implicit current event loop and we want to be compatible with them". But I think reverting to the code before that will always create a new event loop (calling IOLoop()) and set it as the current one.

Yes, this is confusing, and my brief comments left me struggling to piece together what's going on. But the key is to look at [GetNewIOLoopTest](https://github.com/tornadoweb/tornado/blob/9e9043e3cbf4605b272f74ede2c8e6a4d71049c1/tornado/test/testing_test.py#L345). Reverting to the older code will call `self.get_new_ioloop()` every time, which will normally create a new event loop and make it current. But that method can be overridden, and sometimes must be, to allow some surrounding framework (like pytest-asyncio) to initialize an event loop that will be used in the test. 

As long as reverting to the previous version of the testing code passes GetNewIOLoopTest, I think that's all we need to do here.

takluyver commented:
I reverted the code in `AsyncTestCase`. Then there were some ResourceWarnings being emitted in the tests (which caused the tests to fail by writing to stderr). It turned out that removing a mysterious workaround in `test_asyncio_adapter` made them pass again, equally mysteriously, at least on my machine.

takluyver commented:
Ugh, I'm having a hard time understanding how the current machinery (current, make_current, clear_current) is meant to work without a way to check asyncio for a 'current' event loop. `ioloop.current(instance=False)` is kind of a lie, because it can still make a new loop, and even `ioloop.make_current()` may instantiate *another* asyncio event loop while trying to check if it's the current one.

With the more precise deprecation warning in Python 3.11.1, we could convert DeprecationWarning to an error, catch it the same way as RuntimeError, and interpret it as 'no existing event loop'. But in 3.11.0 (and several 3.10.x versions), this will also stop us getting an explicitly set loop (i.e. set by `asyncio.set_event_loop()`).

Or we could just silence the deprecation warnings; either for the tests, or for anything calling tornado's ioloop API. :confused:

takluyver commented:
For now, I've gone back to suppressing the deprecation warning for large parts of the tests.

bdarnell commented:
> If AsyncTestCase did its setup in terms of asyncio.set_event_loop instead of IOLoop.make_current, does that simplify things?

This does appear to work, except that it breaks the poorly-understood pytest-asyncio interactions. Interestingly, the standard library's `unittest.IsolatedAsyncioTestCase` appears to do the same thing, unconditionally creating a new event loop without the option to pull anything in from a surrounding framework. [pytest-asyncio's README](https://github.com/pytest-dev/pytest-asyncio) has a comment about this that I find confusingly worded:

> Note that test classes subclassing the standard [unittest](https://docs.python.org/3/library/unittest.html) library are not supported. Users are advised to use [unittest.IsolatedAsyncioTestCase](https://docs.python.org/3/library/unittest.html#unittest.IsolatedAsyncioTestCase) or an async framework such as [asynctest](https://asynctest.readthedocs.io/en/latest).

Does that mean to use IsolatedAsyncioTestCase *instead of* pytest-asyncio, or that pytest-asyncio works with IsolatedAsyncioTestCase but not with plain unittest.TestCase? I think it's the former; I don't see any tests or examples of pytest-asyncio being used in conjunction with IsolatedAsyncioTestCase. Given that precedent, I'm inclined to just decide that it's not feasible to combine pytest-asyncio with Tornado's AsyncTestCase in this way. (This may get easier since pytest-asyncio appears to have recently switched to a "strict" mode in which it only activates on annotated test methods instead of running on all of them).

takluyver commented:
> There's no reason for Tornado to have its own make_current and clear_current, so maybe we keep them deprecated and let them do wasteful things like create a new asyncio event loop just to throw it away.

That sounds reasonable to me. Having fewer pieces that deal with mutable global state seems like a good idea. So the suggested replacements would be:

- `make_current()` -> `asyncio.set_event_loop(tornado_loop.asyncio_loop)`
- `clear_current()` -> `asyncio.set_event_loop(None)` (up to the application whether they deal with the warnings you mentioned or ignore them)

> I'd be willing to lose the error check in initialize. And could we just change the default for make_current to True (or False; I'm not sure which would be more compatible with actual usage of this method today) and get rid of the old None behavior?

If we're deprecating the `make_current()` method, presumably we would also go for `make_current=False` as the default for initialize, and deprecate `make_current=True`?

> Does that mean to use IsolatedAsyncioTestCase instead of pytest-asyncio, or that pytest-asyncio works with IsolatedAsyncioTestCase but not with plain unittest.TestCase? I think it's the former;

I read it as the latter. I think they're saying that if you prefer to write your tests unittest style (subclassing a TestCase class) but run them with pytest, then you should subclass from one of those options in your test code, and use the pytest-asyncio plugin to run them. This fits with https://github.com/pytest-dev/pytest-asyncio/issues/180 .

> I'm inclined to just decide that it's not feasible to combine pytest-asyncio with Tornado's AsyncTestCase in this way. (This may get easier since pytest-asyncio appears to have recently switched to a "strict" mode in which it only activates on annotated test methods

Indeed, that has made things a lot nicer for compatibility.

So the plan is that `AsyncTestCase` can always create a new IOLoop and set it as the current loop in asyncio? Should it also clear it in tearDown?

takluyver commented:
I've had a go at these changes.

> I'd be willing to lose the error check in initialize. And could we just change the default for make_current to True (or False

I figured that if we were deprecating `IOLoop(make_current=True)` anyway, we may as well leave the behaviour as similar as possible, so I didn't remove the error check.

takluyver commented on tornado/ioloop.py:
```diff
@@ -145,12 +144,10 @@ async def main():
+    .. versionchanged:: 6.3
+       ``make_current=False`` is now the default when creating an IOLoop -
+       previously the default was to make the event loop current if there wasn't
+       already a current one. Passing ``make_current=True`` is deprecated.
```
I just noticed that `make_current=True` was already deprecated in 6.2; I can split this out in the docstring if you like.

takluyver commented on tornado/testing.py:
```diff
@@ -201,41 +190,12 @@ def setUp(self) -> None:
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore", DeprecationWarning)
-            self.io_loop.make_current()
+            self._prev_aio_loop = asyncio.get_event_loop()
```
I think this maintains compatibility with what pytest-asyncio does? At least, it passes the test that appears to be checking this.

It will create an unnecessary event loop when there isn't already a current one. But after the first test has done this, there will be a current event loop, so subsequent tests will only create and tear down the one they're using.

bdarnell replied:
This is sneaky - it works in tornado's test suite because we configure deprecation warnings to be errors. But in a default configuration, it will log a warning and there's no good way to suppress it. (This makes me think that we probably need a CI configuration that doesn't have warnings-as-errors, just like we have one that runs in -O mode to ensure we don't rely on the assert statement which can be disabled). 

If the pytest-asyncio integration were valuable enough I think we could figure out a way to keep it and deal with the warning, but given the precedent of IsolatedAsyncioTestCase I think we're better off just simplifying it. The person who originally asked for this feature in #2324 says they no longer need it, and I suspect that the reason they needed it in the first place was something idiosyncratic about their application rather than something broadly applicable to all pytest users.

takluyver replied:
In this `setUp` method, we're specifically silencing those deprecation warnings (if triggered by tornado code). As I understand the filtering system, this should hide them for anyone using AsyncTestCase, regardless of what options they set in their own code.

However, if it's no longer needed now, we can simplify the code to just clear the event loop, without storing the previous one. I think that means taking out `GetNewIOLoopTest` as well.

bdarnell commented:
> So the suggested replacements would be:

>    make_current() -> asyncio.set_event_loop(tornado_loop.asyncio_loop)

Yes, that's the most direct translation, although even better would be to make the asyncio loop the top-level object instead of reaching into the IOLoop to get it.

>    clear_current() -> asyncio.set_event_loop(None) (up to the application whether they deal with the warnings you mentioned or ignore them)

Yes. One other caveat is that in asyncio, you can't completely reset the state of the global event loop. The implicit event loop creation only works if set_event_loop has never been called (this is why you needed to delete a call to set_event_loop(None) earlier in this PR). Setting it to None doesn't bring back the default behavior, so in some cases it might be more compatible to use `asyncio.set_event_loop(asyncio.new_event_loop)` instead. 

> If we're deprecating the make_current() method, presumably we would also go for make_current=False as the default for initialize, and deprecate make_current=True?

We want to get away from all the "current" methods as near-equivalents for set_event_loop. But in the case of the IOLoop constructor, we create a new event loop, and the question is whether we call `set_event_loop` on it. I think that most of the time, that's going to be what you want. 

In any case, I think the IOLoop constructor is essentially soft-deprecated - no new code should use it, although it would be too disruptive to remove it (of course we must have *some* constructor, so by "remove" I mean make it for internal use of `IOLoop.current()` only). Given that, we want to make the smallest changes we can while working within the constraints of asyncio. The conditional behavior of `make_current=None` is difficult to preserve, but whether it should be replaced by True or False depends on whether there is usually an active event loop when an `IOLoop` is constructed. I think there's usually not, which is why I think `set_event_loop` is probably the right thing to do. 

> So the plan is that AsyncTestCase can always create a new IOLoop and set it as the current loop in asyncio? Should it also clear it in tearDown?

Yes. We want to follow IsolatedAsyncioTestCase in this respect, and it does set_event_loop(None) in tearDown.

takluyver commented:
I'm conscious that you might be spending more time explaining to me what changes should happen than it would take you to make them yourself. Also, I think the timezones mean we're usually taking a day for each exchange of messages.

If you prefer to make the changes directly, go for it. I've ticked the checkbox that lets you push to my branch, or of course you can make a new branch. I started this because it seemed like a relatively straightforward piece of busywork that I could take care of for you. :slightly_smiling_face:

bdarnell commented:
> I'm conscious that you might be spending more time explaining to me what changes should happen than it would take you to make them yourself. Also, I think the timezones mean we're usually taking a day for each exchange of messages.

I'm also conscious of that possibility, but I don't think it's been a problem yet. I appreciate your help here even though it didn't turn out to be a "straightforward piece of busywork". However, now that we're at the finish line I think I'll take over to do the last few cleanups (including a rebase of the commit history). I'm going to remove the silence of deprecation warnings in AsyncTestCase.setUp, deprecate get_new_ioloop, and add a deprecation warning in clear_current.

bdarnell commented:
Actually, the commit history looks pretty decent and rebasing anything looks like more trouble than it's worth (and github is not letting me push to your branch), so I'll merge this and do a quick followup PR for the last couple of changes. Thanks again!

takluyver commented:
Thank you!
